### PR TITLE
Implement insert operation of RB tree

### PR DESCRIPTION
**Pull request by aartdem**
No testing was done

### CODE REVIEW

File: lib/src/main/kotlin/bstrees/RBTree.kt [hunk 1]
```diff
@@ -7,8 +7,17 @@ class RBTree<T : Comparable<T>> : SelfBalancingBST<T, RBNode<T>>() {
     override fun createNewNode(data: T) = RBNode(data)
     override val balancer = RBBalancer<T>()
 
+    /**
+     * Insert data in tree.
+     *
+     * @param data the type of data (should be comparable)
+     * @return Nothing
+     */
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/RBTree.kt]: I don't see any point in using `@param`, `@return` statements here.
I suggest you rewrite the documentation comment as following
```suggestion
    /** Inserts new node with [data] as its value in the tree */
```

File: lib/src/main/kotlin/bstrees/RBTree.kt [hunk 2]
```diff
@@ -7,8 +7,17 @@ class RBTree<T : Comparable<T>> : SelfBalancingBST<T, RBNode<T>>() {
...
     override fun insert(data: T) {
-        TODO("Not yet implemented")
+        // insert like in SimpleBST and paint the new Node red
+        val currentNode = super.insertNode(data)
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/RBTree.kt]: why `super.`?

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 3]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
...
 class RBBalancer<T : Comparable<T>> : TreeBalancer<T, RBNode<T>> {
-    override fun balanceAfterInsertion(node: RBNode<T>): RBNode<T> {
-        TODO("Not yet implemented")
+    private fun color(node: RBNode<T>?) = node?.color ?: RBNode.Color.Black
+
+    private fun isRed(node: RBNode<T>?) = color(node) == RBNode.Color.Red
+
+    private fun isBlack(node: RBNode<T>?) = !isRed(node)
+
+    private fun getParent(node: RBNode<T>) = node.parent
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: What's the point of `getParent`?
It's basically the same as just writing `node.parent`

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 4]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
...
+        return if (parent.left == node) parent.right
+        else parent.left
+    }
+
+    private fun getUncle(node: RBNode<T>): RBNode<T>? {
+        val parent = getParent(node) ?: return null
+        return getSibling(parent)
+    }
+
+    private fun getGrandParent(node: RBNode<T>): RBNode<T>? = node.parent?.parent
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: Same as `getParent`. What's the point of the function? It's used exactly once in the code

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 5]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
...
+
+    private fun rotateRight(node: RBNode<T>): RBNode<T> {
+        val parent = getParent(node)
+        val wasChild = node.left
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt:23]: I suggest you add null check here.
If `node.left` is null then throw `IllegalArgumentException`

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 6]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
...
+
+        if (parent?.left == node) parent.left = wasChild else parent?.right = wasChild
+        wasChild.parent = parent
+
+        return wasChild
+    }
+
+    /**
+     * Returns new tree root (if it was change)
+     */
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: One-line doc comment
```suggestion
    /** Returns new tree root */
```

I also suggest you rewrite the comment as it doesn't reflect what the function actually does

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 7]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
...
+            val parent = getParent(currentNode)
+            if (parent == null) {
+                currentNode.flipColor()
+                break
+            }
+
+            val uncle = getUncle(currentNode)
+            val grandParent = getGrandParent(currentNode)
+            if (isRed(uncle)) {
+                // there parent, uncle and grandParent are not null
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]:
```suggestion
                // here parent, uncle and grandParent can not be null
```

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 8]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
 import bstrees.nodes.RBNode
 
 class RBBalancer<T : Comparable<T>> : TreeBalancer<T, RBNode<T>> {
-    override fun balanceAfterInsertion(node: RBNode<T>): RBNode<T> {
-        TODO("Not yet implemented")
+    private fun color(node: RBNode<T>?) = node?.color ?: RBNode.Color.Black
+
+    private fun isRed(node: RBNode<T>?) = color(node) == RBNode.Color.Red
+
+    private fun isBlack(node: RBNode<T>?) = !isRed(node)
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: image

**Review by therain7 — CHANGES_REQUESTED** [review]
Fails with `NullPointerException` when trying to
```kotlin
tree.insert(1)
tree.insert(2)

```

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt (hunk 8 above)
Comment (aartdem) [reply lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: This function may be needed further

Comment (therain7) [reply lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: ok

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 9]
```diff
@@ -3,8 +3,86 @@ package bstrees.balancers
 import bstrees.nodes.RBNode
 
 class RBBalancer<T : Comparable<T>> : TreeBalancer<T, RBNode<T>> {
-    override fun balanceAfterInsertion(node: RBNode<T>): RBNode<T> {
-        TODO("Not yet implemented")
+    private fun color(node: RBNode<T>?) = node?.color ?: RBNode.Color.Black
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: why not `getColor`?

File: lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt [hunk 10]
```diff
@@ -3,8 +3,79 @@ package bstrees.balancers
...
+    }
+
+    private fun getUncle(node: RBNode<T>): RBNode<T>? {
+        val parent = node.parent ?: return null
+        return getSibling(parent)
+    }
+
+    private fun rotateRight(node: RBNode<T>): RBNode<T> {
+        val parent = node.parent
+        val wasChild = node.left ?: throw IllegalArgumentException("Incorrect argument for function rotateRight()")
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/balancers/RBBalancer.kt]: Description passed to the exception doesn't tell anything about the problem.
It should describe why exactly the argument is incorrect

**Review by therain7 — APPROVED** [review]

**Pull request merged**